File: server/src/main/kotlin/com/heerkirov/hedge/server/dto/res/Trash.kt
package com.heerkirov.hedge.server.dto.res

import com.heerkirov.hedge.server.model.Illust
import java.time.LocalDate
import java.time.Instant

data class TrashedImageRes(val id: Int, val filePath: FilePath,
                           val score: Int?, val favorite: Boolean, val tagme: Illust.Tagme, val source: SourceDataPath?,
                           val orderTime: Instant, val trashedTime: Instant, val remainingTime: Long?)

data class TrashedImageDetailRes(val id: Int, val filePath: FilePath,
                                 val extension: String, val size: Long, val resolutionWidth: Int, val resolutionHeight: Int, val videoDuration: Long,
                                 val topics: List<TopicSimpleRes>, val authors: List<AuthorSimpleRes>, val tags: List<TagSimpleRes>,
                                 val collection: IllustParent?, val books: List<BookSimpleRes>, val folders: List<FolderSimpleRes>, val associates: List<IllustSimpleRes>,
                                 val description: String, val score: Int?, val favorite: Boolean, val tagme: Illust.Tagme, val source: SourceDataPath?,
                                 val partitionTime: LocalDate, val orderTime: Instant,
                                 val createTime: Instant, val updateTime: Instant,
                                 val trashedTime: Instant, val remainingTime: Long?)

File: frontend/src/functions/http-client/api/trash.ts
import { date, datetime, LocalDate, LocalDateTime } from "@/utils/datetime"
import { NotFound } from "../exceptions"
import { HttpInstance, Response } from "../instance"
import { FilePath, LimitAndOffsetFilter, ListResult, mapFromOrderList, OrderList, SimpleAuthor, SimpleTag, SimpleTopic, SourceDataPath } from "./all"
import { SimpleCollection, SimpleIllust, Tagme } from "./illust"
import { SimpleFolder } from "./folder"
import { SimpleBook } from "./book"

export function createTrashEndpoint(http: HttpInstance): TrashEndpoint {
    return {
        list: http.createQueryRequest("/api/trashes", "GET", { 
            parseQuery: mapFromTrashFilter, 
            parseResponse: ({ total, result }: ListResult<any>) => ({total, result: result.map(mapToTrashedImage)}) 
        }),
        get: http.createPathRequest(id => `/api/trashes/${id}`, "GET", { parseResponse: mapToDetailTrashedImage }),
        delete: http.createDataRequest("/api/trashes/delete", "POST"),
        restore: http.createDataRequest("/api/trashes/restore", "POST")
    }
}

function mapFromTrashFilter(filter: TrashFilter): any {
    return {
        ...filter,
        order: mapFromOrderList(filter.order)
    }
}

function mapToTrashedImage(data: any): TrashedImage {
    return {
        ...data,
        trashedTime: datetime.of(<string>data["trashedTime"]),
        orderTime: datetime.of(<string>data["orderTime"]),
    }
}

function mapToDetailTrashedImage(data: any): DetailTrashedImage {
    return {
        ...data,
        partitionTime: date.of(<string>data["partitionTime"]),
        orderTime: datetime.of(<string>data["orderTime"]),
        createTime: datetime.of(<string>data["createTime"]),
        updateTime: datetime.of(<string>data["updateTime"]),
        trashedTime: datetime.of(<string>data["trashedTime"]),
    }
}

/**
 * 已删除项目。
 */
export interface TrashEndpoint {
    /**
     * 查询所有已删除的项目。
     */
    list(filter: TrashFilter): Promise<Response<ListResult<TrashedImage>>>
    /**
     * 查看已删除项目。
     * @exception NOT_FOUND
     */
    get(id: number): Promise<Response<DetailTrashedImage, NotFound>>
    /**
     * 彻底删除已删除项目。
     */
    delete(imageIds: number[]): Promise<Response<unknown>>
    /**
     * 还原已删除项目。
     */
    restore(imageids: number[]): Promise<Response<unknown>>
}

export interface TrashedImage {
    id: number
    filePath: FilePath
    score: number | null
    favorite: boolean
    tagme: Tagme[]
    source: SourceDataPath | null
    orderTime: LocalDateTime
    trashedTime: LocalDateTime
    remainingTime: number | null
}

export interface DetailTrashedImage extends TrashedImage {
    extension: string
    size: number
    resolutionHeight: number
    resolutionWidth: number
    videoDuration: number
    topics: SimpleTopic[]
    authors: SimpleAuthor[]
    tags: SimpleTag[]
    collection: SimpleCollection | null
    books: SimpleBook[]
    folders: SimpleFolder[]
    associates: SimpleIllust[]
    description: string
    partitionTime: LocalDate
    createTime: LocalDateTime
    updateTime: LocalDateTime
}

export type TrashFilter = TrashQueryFilter & LimitAndOffsetFilter

export interface TrashQueryFilter {
    order?: OrderList<"id" | "trashedTime" | "orderTime">
}
